File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/domain/java/PaymentService.java
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.domain.java;

// Service class that handles domain logic.
public class PaymentService {

    public String describeResult(PaymentResult result) {
        // Pattern matching with sealed types allows Java to handle
        // different result types in a controlled way.
        if (result instanceof PaymentSuccess success) {
            return "Payment completed: " + success.transactionId();
        }

        if (result instanceof PaymentFailure failure) {
            return "Payment failed: " + failure.reason();
        }

        throw new IllegalStateException("Unknown payment result");
    }
}

File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/domain/kotlin/PaymentService.kt
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.domain.kotlin

// Service class that handles domain logic.
class PaymentService {

    fun describeResult(result: PaymentResult): String {
        // when expression works very well with sealed classes.
        // Kotlin knows all possible subclasses of PaymentResult.
        return when (result) {
            is PaymentResult.Success ->
                "Payment completed: ${result.transactionId}"

            is PaymentResult.Failure ->
                "Payment failed: ${result.reason}"
        }
    }
}